# Task 1 :      
Check the code from this github link https://github.com/google-research/bleurtLinks to an external site. Run the code on the data given in this repository.

In [ ]:
import sys
print("User Current Version:-", sys.version)

User Current Version:- 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0]


## Clone the github .

In [ ]:
! git clone https://github.com/google-research/bleurt.git

Cloning into 'bleurt'...
remote: Enumerating objects: 134, done.
remote: Counting objects: 100% (18/18), done.
remote: Compressing objects: 100% (18/18), done.
remote: Total 134 (delta 0), reused 17 (delta 0), pack-reused 116
Receiving objects: 100% (134/134), 31.28 MiB | 32.71 MiB/s, done.
Resolving deltas: 100% (49/49), done.


In [ ]:
cd /content/bleurt

/content/bleurt


## Install the BLEURT library

In [ ]:
pip install .

In [ ]:
pip list

## Running Unit Tests

In [ ]:

import unittest
import bleurt

In [ ]:
!python -m unittest bleurt.score_test

2023-11-18 15:35:56.236741: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-11-18 15:35:56.236817: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-11-18 15:35:56.236877: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-11-18 15:35:57.380501: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
2023-11-18 15:36:00.083415: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:47] Overriding orig_value setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
.....s..
-------------

In [ ]:
!python -m unittest bleurt.score_not_eager_test

2023-11-18 15:36:17.000108: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-11-18 15:36:17.000167: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-11-18 15:36:17.000201: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-11-18 15:36:18.269395: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
2023-11-18 15:36:21.985427: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:47] Overriding orig_value setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
.s.
------------------

In [ ]:
!python -m unittest bleurt.score_files_test

2023-11-18 15:37:58.542496: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-11-18 15:37:58.542545: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-11-18 15:37:58.542588: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-11-18 15:37:59.670344: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
2023-11-18 15:38:02.258967: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:47] Overriding orig_value setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
........s..
----------

## Using BLEURT for Scoring

Once the tests pass, we can use BLEURT for scoring. Import the BLEURT scorer and load the pre-trained model.

## Download BLEURT Base Checkpoint

In [ ]:
!wget https://storage.googleapis.com/bleurt-oss-21/BLEURT-20.zip
!unzip BLEURT-20.zip


--2023-11-18 15:57:27--  https://storage.googleapis.com/bleurt-oss-21/BLEURT-20.zip
Resolving storage.googleapis.com (storage.googleapis.com)... 209.85.146.207, 209.85.147.207, 142.250.125.207, ...
Connecting to storage.googleapis.com (storage.googleapis.com)|209.85.146.207|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2140294207 (2.0G) [application/octet-stream]
Saving to: ‘BLEURT-20.zip’

BLEURT-20.zip       100%[===================>]   1.99G   192MB/s    in 13s     

2023-11-18 15:57:40 (159 MB/s) - ‘BLEURT-20.zip’ saved [2140294207/2140294207]

Archive:  BLEURT-20.zip
   creating: BLEURT-20/
  inflating: BLEURT-20/bert_config.json  
  inflating: BLEURT-20/saved_model.pb  
   creating: BLEURT-20/variables/
  inflating: BLEURT-20/variables/variables.index  
  inflating: BLEURT-20/variables/variables.data-00000-of-00001  
  inflating: BLEURT-20/sent_piece.vocab  
  inflating: BLEURT-20/bleurt_config.json  
  inflating: BLEURT-20/sent_piece.model  


## Run BLEURT Scoring

  We can run BLEURT scoring on a sample set of sentences.

    -candidate_file: Specifies the file containing candidate sentences.
    -reference_file: Specifies the file containing reference sentences.
    -bleurt_checkpoint: Specifies the BLEURT checkpoint or model to use (BLEURT-20 in this case).


* Candidates :      



        1.  Bud Powell was a legendary pianist.
        2.  Bud Powell was a historical piano player.
        3.  Bud Powell was a new yorker.
        4.  Bud a day keeps the doctor away.


* References :       

      1. Bud Powell was a legendary painist.
      2. Bud Powell was a legendary painist.
      3.Bud Powell was a legendary painist.
      4.Bud Powell was a legendary painist.







In [ ]:
!python -m bleurt.score_files \
  -candidate_file=bleurt/test_data/candidates \
  -reference_file=bleurt/test_data/references \
  -bleurt_checkpoint=BLEURT-20


# So  BLEURT scoring process has been completed successfully!

Output ▶



*   Scores ▶
        

          *   0.9777958393096924
          *   0.7689998149871826
          *   0.4400966763496399
          *   0.16677340865135193

These numerical values represent the BLEURT scores for the corresponding candidate-reference pairs. The higher the score, the better the alignment or quality of the candidate sentence with the reference sentences.











## Python API

In [ ]:
from bleurt import score

In [ ]:
checkpoint = "bleurt/test_checkpoint"
references = ["This is a test."]
candidates = ["This is the test."]

scorer = score.BleurtScorer(checkpoint)
scores = scorer.score(references=references, candidates=candidates)
assert isinstance(scores, list) and len(scores) == 1
print(scores)

[0.7463243007659912]


## Tensorflow API

In [ ]:
import tensorflow as tf
from bleurt import score

In [ ]:
references = tf.constant(["This is a test."])
candidates = tf.constant(["This is the test."])

bleurt_ops = score.create_bleurt_ops()
bleurt_out = bleurt_ops(references=references, candidates=candidates)




In [ ]:
assert bleurt_out["predictions"].shape == (1,)

print(bleurt_out["predictions"])

tf.Tensor([0.7463243], shape=(1,), dtype=float32)


This output is a single BLEURT score . The exact value of the score will depend on the BLEURT model and its assessment of how well the candidate sentence ("This is the test.") aligns with the reference sentence ("This is a test.").

In [ ]:
# numpy version
print('Score is :',bleurt_out["predictions"].numpy())

Score is : [0.7463243]


## Speeding Up BLEURT

     Download the Distilled Model

In [ ]:
!wget https://storage.googleapis.com/bleurt-oss-21/BLEURT-20-D12.zip
!unzip BLEURT-20-D12.zip


--2023-11-18 16:27:53--  https://storage.googleapis.com/bleurt-oss-21/BLEURT-20-D12.zip
Resolving storage.googleapis.com (storage.googleapis.com)... 74.125.124.207, 172.217.212.207, 172.217.214.207, ...
Connecting to storage.googleapis.com (storage.googleapis.com)|74.125.124.207|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 625304855 (596M) [application/octet-stream]
Saving to: ‘BLEURT-20-D12.zip’

BLEURT-20-D12.zip   100%[===================>] 596.34M  80.4MB/s    in 10s     

2023-11-18 16:28:03 (58.5 MB/s) - ‘BLEURT-20-D12.zip’ saved [625304855/625304855]

Archive:  BLEURT-20-D12.zip
   creating: BLEURT-20-D12/
  inflating: BLEURT-20-D12/bert_config.json  
  inflating: BLEURT-20-D12/saved_model.pb  
   creating: BLEURT-20-D12/variables/
  inflating: BLEURT-20-D12/variables/variables.index  
  inflating: BLEURT-20-D12/variables/variables.data-00000-of-00001  
  inflating: BLEURT-20-D12/sent_piece.vocab  
  inflating: BLEURT-20-D12/bleurt_config.json  
  in

## Run BLEURT with Optimizations

In [ ]:
%%time
!python -m bleurt.score_files \
  -candidate_file=/content/bleurt/bleurt/test_data/candidates \
  -reference_file=/content/bleurt/bleurt/test_data/references \
  -bleurt_batch_size=100 \
  -batch_same_length=True \
  -bleurt_checkpoint=BLEURT-20-D12


2023-11-18 16:31:32.146052: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-11-18 16:31:32.146109: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-11-18 16:31:32.146141: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-11-18 16:31:33.212173: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
INFO:tensorflow:Running BLEURT scoring.
I1118 16:31:34.733731 137748863459328 score_files.py:168] Running BLEURT scoring.
W1118 16:31:34.733995 137748863459328 score_files.py:118] Enabling same length batching. BEWARE: this is an experime

BLEURT scores are :

    0.9525628685951233
    0.7156853079795837
    0.4253561496734619
    0.23305043578147888  

It took CPU times: user 86.8 ms, sys: 15.5 ms, total: 102 ms , Wall time: 11.7 s

# Task 2 :
**Point out at least three categories that this metric fails to capture, and give examples for each category. Then propose concrete solutions on addressing the challenges.**  

## Category 1 :    
### Detection of Sarcasm or Irony

In [ ]:
from bleurt import score

In [ ]:
# We will use python API
checkpoint = "bleurt/test_checkpoint"
references = ["The movie was captivating and kept me on the edge of my seat."]
candidates = ["The movie was so good that I slept through the whole thing."]

scorer = score.BleurtScorer(checkpoint)
scores = scorer.score(references=references, candidates=candidates)
assert isinstance(scores, list) and len(scores) == 1
print(scores)

[-1.0145339965820312]


We see that score is : -1.0145339965820312  . That means both sentence are totally uncorellated where both meaning are same .

***
**BLEURT  metrics struggle with sarcasm or irony**.

## Category 2 :    
### Cultural and Idiomatic Expressions

In [ ]:
# We will use python API
# Example
checkpoint = "bleurt/test_checkpoint"
references = ["He spilled the beans about the surprise party."]
candidates = ["He revealed the secret about the surprise party."]

scorer = score.BleurtScorer(checkpoint)
scores = scorer.score(references=references, candidates=candidates)
assert isinstance(scores, list) and len(scores) == 1
print(scores)

[-0.5376939177513123]


We can see the BLEURT score is -0.5376939177513123 , which mean uncorrelated .
***
But the idiom "spilled the beans" means to reveal a secret. While the candidate sentence conveys the same meaning, BLEURT might not recognize the equivalence due to the idiomatic expression.
***

**BLEURT might struggle with sentences that contain culturally specific references or idiomatic expressions.**

## Category 3 :    
### Subtlety in Politeness or Formality

In [ ]:
# We will use python API
# Example
checkpoint = "bleurt/test_checkpoint"
references = ["Could you possibly open the window?"]
candidates = ["Open the window, will you?"]

scorer = score.BleurtScorer(checkpoint)
scores = scorer.score(references=references, candidates=candidates)
assert isinstance(scores, list) and len(scores) == 1
print(scores)

[-0.15784744918346405]


We can see the BLEURT score is -0.15784744918346405 , which mean uncorrelated .
***
Although both sentences are requests to open the window, the tone and politeness level differ significantly. BLEURT might not capture this difference effectively, as it's more about social context than linguistic content.

***
**BLEURT may not effectively capture these subtleties, especially in languages where formality is grammatically encoded.**

## Probable solutions to handle these problems .

**Answer** :

*  Enhanced Training Data:
     
     Objective :  To improve the model's ability to understand and evaluate a wide range of linguistic expressions, especially those that are culturally or contextually specific.

     Implementation : We have to gather the data from variety sources from different styles, regions, and cultural backgrounds.we have to includes those datasets with high instances of sarcasm, idioms, and figurative language.

* Contextual Understanding :    

    To develop models that can take into account the broader context in which language is used, not just the literal content of sentences.

   we have to develop algorithms that can consider the wider context of the conversation or text. This might include understanding the preceding dialogue, the overall tone, or the specific situation in which the language is used.

   Also we have to employ advance deep learning techniques .

*  Human-in-the-Loop :      
   
   This is the most important idea like To leverage human judgment in conjunction with automated metrics for a more accurate and nuanced evaluation of language.

   Generally there are 2 type of implementations are there :    



            1.   Hybrid Evaluation Systems
            2.   Feedback Mechanism
   Both are basically those type of implementation where  automated evaluations  and human feedback will be used to continually .

   ***

   By implementing these strategies, evaluation metrics like BLEURT can be significantly improved to handle the intricacies of human language..



   
